Document::cssParserContext() should return a reference

geoffreygaren opened the pull request:
#### 6a9292ad54735f7eb30651960723c3de9552dac3
<pre>
Document::cssParserContext() should return a reference
<a href="https://bugs.webkit.org/show_bug.cgi?id=291761">https://bugs.webkit.org/show_bug.cgi?id=291761</a>
<a href="https://rdar.apple.com/149561463">rdar://149561463</a>

Reviewed by Sam Weinig.

This removes 500K string refcounts from Speedometer, primarily by enabling
InlineCSSStyleProperties::cssParserContext() to reuse Document::cssParserContext()
in the common case.

That&apos;s 500K / 34MM, so this patch is not a measurable speedup on its own; but
the reduction is countable, and this is also an opportunity to land
OptionalOrReference&lt;T&gt; with a real use case, to complement ValueOrReference&lt;T&gt;.

* Source/WTF/wtf/OptionalOrReference.h: Copied from Source/WTF/wtf/ValueOrReference.h.
(WTF::OptionalOrReference::OptionalOrReference): New helper type that&apos;s just
like ValueOrReference&lt;T&gt; but doesn&apos;t require T() to be efficient.

Canonical link: <a href="https://commits.webkit.org/295371@main">https://commits.webkit.org/295371@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d84b8d865aea69ef4c64c52a2816faf9c8c2cc35

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104856 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24569 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/14990 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/110071 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/55530 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/106896 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/24969 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/33114 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79621 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107862 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/19416 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94629 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59928 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/19163 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/12706 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/54913 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/97537 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/88862 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12753 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/112513 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/103474 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/32021 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/23539 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/88701 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/32385 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90855 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88329 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33216 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10983 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/27335 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17019 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31946 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/37301 "Built successfully") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/127753 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/31738 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/127753 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/35079 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/33297 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WTF/wtf/ValueOrReference.h:
```diff
@@ -73,6 +73,36 @@ template<typename T> class ValueOrReference {
     const T& m_reference;
 };
 
+// Use OptionalOrReference when T does not have an efficient default constructor.
```
I think this could be expanded a bit. Perhaps explaining that it should be used as a return value or something like that.

weinig commented on Source/WebCore/css/query/MediaQueryParser.h:
```diff
@@ -34,18 +34,18 @@ namespace MQ {
 struct MediaProgressProviding;
 
 struct MediaQueryParser : public GenericMediaQueryParser<MediaQueryParser>  {
-    static MediaQueryList parse(const String&, const MediaQueryParserContext&);
```
This is excellent. I totally forgot I added a fixme to replace all MediaQueryParserContext with CSSParserContext a few months back. I bet this gets us pretty close (or all the way) to completely removing MediaQueryParserContext.

weinig reviewed (approved):
Excellent change!

geoffreygaren committed:
Document::cssParserContext() should return a reference
https://bugs.webkit.org/show_bug.cgi?id=291761
rdar://149561463

Reviewed by Sam Weinig.

This removes 500K string refcounts from Speedometer, primarily by enabling
InlineCSSStyleProperties::cssParserContext() to reuse Document::cssParserContext()
in the common case.

That's 500K / 34MM, so this patch is not a measurable speedup on its own; but
the reduction is countable, and this is also an opportunity to land
OptionalOrReference<T> with a real use case, to complement ValueOrReference<T>.

* Source/WTF/wtf/OptionalOrReference.h: Copied from Source/WTF/wtf/ValueOrReference.h.
(WTF::OptionalOrReference::OptionalOrReference): New helper type that's just
like ValueOrReference<T> but doesn't require T() to be efficient.

Canonical link: https://commits.webkit.org/295371@main
